File: js/i18n.js
// ── i18n.js — lightweight key-based localisation engine ─────────────
// No dependencies. Locale modules live in js/locales/.
// Call await initLang() once at bootstrap (top-level await in app.js)
// before any view renders — after that t() is fully synchronous.

let _s = {};   // active locale strings (set by initLang)

// Look up a dot-separated key in _s; substitute {token} placeholders.
// Falls back to the key string itself when a key is missing, so a
// missing translation is always visible rather than silently blank.
export function t(key, vars = {}) {
  const raw = key.split('.').reduce((o, k) => (o && typeof o === 'object' ? o[k] : undefined), _s);
  const str = (raw !== undefined && raw !== null) ? String(raw) : key;
  return Object.entries(vars).reduce((s, [k, v]) => s.replaceAll(`{${k}}`, v), str);
}

// Return the singular or plural form based on count n.
export function plural(n, one, other) { return n === 1 ? one : other; }

// Load a locale by code; if no argument, reads ht_settings.lang from
// localStorage (defaults 'en'). Called with top-level await in app.js.
export async function initLang(lang) {
  if (!lang) {
    try {
      const saved = JSON.parse(localStorage.getItem('ht_settings') || '{}');
      lang = saved.lang || 'en';
    } catch { lang = 'en'; }
  }
  try {
    const mod = await import(`./locales/${lang}.js`);
    _s = mod.default;
  } catch {
    // Locale file failed — fall back to English rather than showing raw keys
    if (lang !== 'en') {
      try {
        const mod = await import('./locales/en.js');
        _s = mod.default;
      } catch { /* leave _s empty; t() will return key strings */ }
    }
  }
}

export function getLang() { return _s._lang ?? 'en'; }

File: flutter_app/lib/core/i18n.dart
// ── i18n.dart — tiny translation engine (port of js/i18n.js) ─────────
// t('namespace.key', {token: value}) with {placeholder} substitution.
// English is the fallback for any key missing from the active locale.
import 'locales/en.dart';
import 'locales/ml.dart';

class I18n {
  String lang = 'en';
  Map<String, Map<String, String>> strings = en;

  void load(String code) {
    lang = code == 'ml' ? 'ml' : 'en';
    strings = lang == 'ml' ? ml : en;
  }
}

final i18n = I18n();

String t(String key, [Map<String, Object?>? vars]) {
  final dot = key.indexOf('.');
  if (dot < 0) return key;
  final ns = key.substring(0, dot);
  final k = key.substring(dot + 1);
  var s = i18n.strings[ns]?[k] ?? en[ns]?[k] ?? key;
  if (vars != null) {
    vars.forEach((token, v) {
      s = s.replaceAll('{$token}', '$v');
    });
  }
  return s;
}

String plural(num n, String one, String other) => n == 1 ? one : other;
